### PR TITLE
Add support for conditional requests

### PR DESCRIPTION
**Pull request by rickb777**

### CODE REVIEW

**Comment by agilestevesmith** [comment]
Thanks @rickb777 we will review this soon

**Comment by duncancrawford** [comment]
@agilestevesmith could you prioritise PRs from @rickb777 as there delivery has a dependency on functionality in this lib

File: src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala [hunk 1]
```diff
@@ -25,6 +25,38 @@ case class UserId(value: String) extends AnyVal
...
+ *
+ * https://tools.ietf.org/html/rfc7232#section-3
+ *
+ * Only ETag preconditions are supported at present.
+ * If-Modified-Since, if-Unmodified-Since and If-Range are not supported.
+ *
+ * The string values should be exactly as provided by earlier responses via ETag
+ * headers. That is, the quote marks and any weak prefix must be used verbatim.
+ */
+case class Precondition(ifMatch: Seq[String] = Seq(),
```
Comment (howyp) [inline src/main/scala/uk/gov/hmrc/play/http/HeaderCarrier.scala]: Perhaps `Precondition` should be a trait, and `ifMatch`/`ifNoneMatch` be implementations of it?

**Comment by howyp** [comment]
@rickb777 There seem to be quite a lot of files that have been auto-formatted - which in some cases makes the code harder to read, and also makes it quite hard to see what functionality has been changed. Would you mind reverting the re-formats where the code hasn't changed?

Adding support for HEAD seems like a good idea to me.

I'm not sure I immediately get the difference between adding a `Precondition` and adding some extra headers into the `HeaderCarrier` - I think the latter would achieve the same thing with existing functionality? This is how the needs of play-partials are implemented, for instance.

I can see the advantage in adding stronger typing for preconditions, but I'm not sure that adding it as a primary parameter to each verb is a great way to go - do we do this for all request headers? This method change will also break binary compatibility for most users, which will be an issue for anyone using http-verbs and also using a library which uses http-verbs, it would be good to avoid this if possible.

**Comment by agilestevesmith** [comment]
Thanks @howyp, I agree with your feedback

@rickb777, thanks as ever for your PRs on this library. We really appreciate your contributions. However, in this particular instance we will decline the PR as:

1. play-partials will give you what you need, see https://github.com/hmrc/play-partials/blob/master/src/main/scala/uk/gov/hmrc/play/partials/HeaderCarrierForPartials.scala#L42.
2. http-verbs cannot take any more breaking changes in the short term as it is the largest diamond dependency in MDTP

Can I also ask that future PRs do not contain autoformatting? Thanks

If you need any help with the play-partials approach, contact @howyp or myself

**Comment by rickb777** [comment]
Hi,

Thanks for the feedback. Fragile binary compatibility is a sore point for Scala, alas.

I have a suggestion. Could we implement this as a major version upgrade? That gets around the risk of binary incompatibility.

On the matter of source code formatting. please do it yourselves. It's frankly not good enough for code rot to set in and not be addressed.

**Comment by agilestevesmith** [comment]
Hi

As I said, we cannot take a major version at the moment as http-verbs is a diamond dependency. The play-partials suggestion from @howyp should give you what you need.

On source code formatting - I'm afraid you are wrong. PRs into HMRC repositories should couple behavioural changes with formatting changes - that coupling makes changes more difficult, not the format of changes. It is for the owners of a repository to format the code according to MDTP guidelines, and for those owners to correct formatting when they deem necessary.

**Comment by rickb777** [comment]
On the latter point, that was what I was saying - it's up to you to format the code, as repo owner. And as a consumer and PR contributor, I expect you to do it.

On the play-partials, we don't use them and won't be needing them. How can that solve this problem?

**Comment by rickb777** [comment]
PS' I'd be quite happy to resubmit a new PR without any extraneous whitespace formatting changes.

**Pull request closed without merging**